Run onstart/onstop when a new song is loaded

SimonPersson opened the pull request:
This PR introduces two changes:
* onstop is run when a new song is loaded, if we were already playing
* onstart is run when a new song is loaded, if we also start playing it

The previous behaviour was that onstop/onstart were only run if the state went from playing => !playing or !playing => playing. The new behaviour is useful for scripts where you want to run a script on every new song. I would argue that there is a brief stop between two songs, so the names still make sense :)

kingosticks commented:
Are you sure about that previous behaviour? Is that the result of a recent change? I'm sure that my onstart script runs for every new song that's played.  I know this as the log file is full of messages saying the command succeeded.

SimonPersson commented:
@kingosticks  - It seems like the scripts run if the song changes due to end of track, but not if you skip a song halfway through.

awiouy commented:
I had to modify the current behaviour, too (https://github.com/awiouy/librespot/commit/6e82353ba271427fe996a6d866bebc70ff96786b).
We should maybe consider add hooks/options:
- an option to not stop the sink
- a hook when session opens/closes
- a hook when sink starts/stops
- a hook when a track is loaded

plietar commented:
Having different behaviours when a song ends and when a song is skipped is definitely wrong.

In general I think it would be better for both to have an `onchange` callback instead. If playback stops (last song is reached and repeat is disabled as well) then `onstop` is called as well.

Additionally, we currently don't pass any information to the caller. I guess it could be useful to do so (someone wants update a display of the current playing song), as a first step torwards #7

plietar commented:
Additionally I think the hooks should be usable from rust code.

Essentially the player should change to be something like:
```rust
enum Event {
    Started {
        track_id: SpotifyId,
    }
    Changed {
        old_track_id: SpotifyId,
        new_track_id: SpotifyId,
    }
    Stopped {
        track_id: SpotifyId,
    }
}

trait EventListener {
    fn on_event(&mut self, event: Event);
}

struct PlayerInternal {
    listener: Option<Box<EventListener>>,
    ...
}

// In various PlayerInternal methods
self.listener.on_event(Event::Started { track_id: track_id });
```

Then a type in `main.rs` would implement the `player::EventListener` trait, by exec'ing whatever is in the configuration, possibly setting environmant variables or doing some string substitution.

Maybe `FnMut` or a channel should be used instead of `player::EventListener`.

michaelherger commented:
FWIW: I'm glad this is coming up, as I did customize the behavior on my end, too. I added an onchange event. in case of a play even, if nothing was playing, I'd run onstart, if something was playing before, onchange.

In addition I call onchange when seeking. And added a volume change handler...

Overall I think there should be a larger review of this. Everybody seems to not be happy what there is, but customizations vary. I like @plietar's suggestion which would give the application more control over how to act under what circumstances (if written in Rust).

SimonPersson commented:
I'm glad I'm not the only one wanting this! I've implemented the interface suggested by @plietar, although over a channel instead of the `EventListener` trait. It seems a bit too easy to shoot yourself in the foot by blocking for too long in the callback otherwise.

I implemented a player state `EndOfTrack`. It's supposed to be for cases where the player is not currently playing because it is waiting for a new track. Previously the `Stopped` state was used for this state, but that makes it impossible to tell wether we should send a `Change` or `Started` event once a new track loads. Perhaps someone could think of a more clever solution here?

plietar commented on src/player_event_handler.rs:
```diff
@@ -0,0 +1,50 @@
+    let status = Command::new(&v.remove(0))
+        .args(&v)
+        .args(args)
+        .status()
```
I would just use `spawn()` and not care about waiting for it.

plietar commented on src/player_event_handler.rs:
```diff
@@ -0,0 +1,50 @@
+    let mut v: Vec<&str> = program.split_whitespace().collect();
+    let status = Command::new(&v.remove(0))
+        .args(&v)
+        .args(args)
```
I think it's better to use environment variables here so the values have names associated. Additionally include a `PLAYER_EVENT=start` etc.. variable so a single program can deal with all events.

You can use a HashMap to pass the arguments in to this function.

plietar commented on src/player_event_handler.rs:
```diff
@@ -0,0 +1,50 @@
+            while let Ok(msg) = receiver.recv() {
+                match msg {
+                    PlayerEvent::Changed { old_track_id, new_track_id } => {
+                        let args = vec![old_track_id.to_base16(), new_track_id.to_base16()];
```
Encoding in base62 is better, as it's the usual "spotify url" format.

SimonPersson replied:
I agree, but perhaps this can wait until `u128` lands on stable? That would make the implementation a lot simpler.

plietar commented on playback/src/player.rs:
```diff
@@ -375,15 +387,14 @@ impl PlayerInternal {
+            Some(ref s) =>
+                match s.send(event.clone()) {
+                    Ok(_) => info!("Sent event {:?} to event listener.", event),
+                    Err(err) => error!("Failed to send event {:?} to listener: {:?}", event, err)
```
Error happens if the receiver is dropped, which is not a big problem. Just ignore the error in that case.

plietar commented on playback/src/player.rs:
```diff
@@ -375,15 +387,14 @@ impl PlayerInternal {
+        match self.config.event_sender {
+            Some(ref s) =>
+                match s.send(event.clone()) {
+                    Ok(_) => info!("Sent event {:?} to event listener.", event),
```
No need for this I think. There's already an `info` for running the program.

plietar commented on src/player_event_handler.rs:
```diff
@@ -0,0 +1,50 @@
+use librespot::playback::config::PlayerEvent;
+
+fn run_program(program: &str, args: Vec<String>) {
+    info!("Running {}", program);
```
Include the arguments in here. Also may be better to print the split version of program, as it may help debugging eg. someone expecting librespot to properly handle quotes/escaped whitespaces.

As a side note, we'd want something like [shellwords](https://docs.rs/shellwords) for smarter splitting, but this does not have to be in this PR.

plietar commented on playback/src/player.rs:
```diff
@@ -375,15 +387,14 @@ impl PlayerInternal {
-        if let Some(ref program) = self.config.onstop {
-            run_program(program)
+    fn send_event(&mut self, event: PlayerEvent) {
+        match self.config.event_sender {
```
Can use an `if let Some(ref s) = ...`

plietar commented on playback/src/player.rs:
```diff
@@ -274,9 +278,15 @@ impl PlayerInternal {
-                self.run_onstop();
 
-                let old_state = mem::replace(&mut self.state, PlayerState::Stopped);
+                let new_state = match self.state {
```
`handle_packet` should only be called from a `Playing` state, so you can `panic!` on other cases.

Additionally you can change `signal_end_of_track` into `playing_to_end_of_track(&mut self)`, which takes care of switching state and sending the signal.

plietar commented on playback/src/player.rs:
```diff
@@ -293,32 +303,35 @@ impl PlayerInternal {
                                 decoder: decoder,
                                 end_of_track: end_of_track,
                             };
+                            self.send_event(PlayerEvent::Stopped { track_id });
```
There should maybe be a `Changed` event as well here. I'm not really sure about the ordering

plietar commented on playback/src/config.rs:
```diff
@@ -25,19 +27,33 @@ impl Default for Bitrate {
 }
 
+#[derive(Debug, Clone)]
+pub enum PlayerEvent {
```
This type should be in `player.rs`

SimonPersson commented:
Thanks for the feedback @plietar! I pushed some new commits adressing all the comments.

sashahilton00 commented:
The event hooks notifier is on my list of things to look into. I've been giving it a bit of thought, as much like the audio sink backend, we will most likely want to support multiple means of publishing metadata, though this will be something that librespotd should handle. For librespot, It would be good to have a fairly verbose event hook system that notifies of pretty much everything (play, pause, seek, stop, repeat, shuffle, etc.) and then the client app can filter out what should be displayed and what actions should be triggered on what events.

Once that's implemented, my vote would also be to move the onstart/onstop/onchange logic currently in librespot into librespotd, as it doesn't feel like something that should be in the core library long term, more like something that is a stopgap whilst we haven't got proper event notification hooks written.

plietar commented on playback/src/player.rs:
```diff
@@ -1,10 +1,10 @@
 use futures::sync::oneshot;
 use futures::{future, Future};
+use futures::sync::mpsc::{UnboundedReceiver, UnboundedSender, unbounded};
```
Given that both `std::sync::mpsc` and `futures::sync::mpsc` are used in this file, I'd recommend using the fully qualified path everytime.

plietar commented on playback/src/player.rs:
```diff
@@ -145,33 +165,44 @@ impl PlayerState {
+    fn playing_to_end_of_track(&mut self) {
+        use self::PlayerState::*;
+        match *self {
+            Paused { track_id, .. } |
```
Only do `Playing`, not `Paused`.

plietar commented on playback/src/player.rs:
```diff
@@ -145,33 +165,44 @@ impl PlayerState {
+            Paused { track_id, .. } |
+            Playing { track_id, .. } => {
+                let old_state = mem::replace(self, EndOfTrack { track_id });
+                old_state.send_end_of_track();
```
You should just merge the `send_end_of_track` function into here, so you only have to match once.
You'll probably have to do the `::std::mem::replace(self, Invalid)` "dance" in other to get the `end_of_track` by value. (See `paused_to_playing` for example).

SimonPersson replied:
Ahh, I was wondering if there was a more elegant way to do that! It didn't occur to me to pattern match on the result of  `replace(self, Invalid)`.

plietar replied:
Hopefully rust will get NLL soon and we can get rid of all this awkward `Invalid` stuff

SimonPersson commented:
I've pushed fixes to the new comments.

plietar commented on playback/src/player.rs:
```diff
@@ -145,33 +165,32 @@ impl PlayerState {
+        match mem::replace(self, Invalid) {
+            Playing { track_id, end_of_track, ..} => {
+                end_of_track.send(());
+                let old_state = mem::replace(self, EndOfTrack { track_id });
```
This can just be `*self = EndOfTrack { ... }`

SimonPersson replied:
Fixed!

sashahilton00 commented:
@SimonPersson if you could resolve the conflicts and push the changes, this should be good to merge.

sashahilton00 commented:
Have resolved the conflicts and fixed the warning, will merge once Travis gives the go ahead.
